flang/lib/Lower/ConvertExpr.cpp

```diff
@@ -1240,15 +1274,18 @@ class ScalarExprLowering {
           fir::emitFatalError(loc, "triplet slice should be handled in genarr");
         }
       }
-      auto ty = genSubType(base.getType(), args.size());
+
+      auto ty = box.match(
+          [&](const fir::IrBoxValue &x) -> mlir::Type { return x.getEleTy(); },
+          [&](const auto &x) -> mlir::Type {
+            return genSubType(base.getType(), args.size());
+          });
       ty = builder.getRefType(ty);
       auto addr = builder.create<fir::CoordinateOp>(loc, ty, base, args);
-      // FIXME: return may not be a scalar.
-      return box.match(
-          [&](const fir::CharArrayBoxValue &x) -> fir::ExtendedValue {
-            return fir::CharBoxValue{addr, x.getLen()};
-          },
-          [&](const auto &) -> fir::ExtendedValue { return addr; });
+      // FIXME: return may not be a scalar. Is this still true given how
+      // array expression are handled ? If not genSubType is also not
+      // needed.
```

Comment:
It should be the case that if this value has rank > 0, then this code is not called. We should probably add an assert just to be sure though.

Reply:
simplified the code and added an assert.